### PR TITLE
Retain ServiceAccount secrets on reapply; CI tweaks

### PR DESCRIPTION
**Pull request by dhaiducek**
The primary goal is to fix the ServiceAccount reconcile that is generating an excessive number of Secrets. As it turns out, this appears to only be relevant for Kubernetes v1.22 and older since newer Kubernetes use tokens instead of secrets.

ref: https://issues.redhat.com/browse/ACM-2049

In the process of working on that issue, I also ran into some CI issues and some other things that I sought to address:
- Implement version files (Because parsing a Makefile isn't very elegant)
- Update workflows to use `release-*` branches 
- Makefile tweaks:
  - Conditionally install `bats`
  - Implement `clean-manifests` and update/replace `sed` with Kustomize patches
  - Set `docker-build` architecture since `image-builder` is only built for `linux/amd64`
  - Add `-v` for E2E test runs
  - Add restoring kustomization to `clean` Make target
  - Shorten output for the SVG base64 in the CSV
- Fix customizing the Docker registry port (If the current registry has a different port, the docker push won't work. This adds some logic to make re-running it work consistently, even if the desired configuration changes.)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -3,10 +3,10 @@
 # To re-generate a bundle for another specific version without changing the standard setup, you can:
 # - use the VERSION as arg of the bundle target (e.g make bundle VERSION=0.0.2)
 # - use environment variables to overwrite this value (e.g export VERSION=0.0.2)
-VERSION ?= 3.11.1
+VERSION ?= $(shell cat VERSION)
```
Comment (yiraeChristineKim) [inline Makefile:6]: Can we add a default? `:=3.11.1`

File: Makefile [hunk 2]
```diff
@@ -3,10 +3,10 @@
...
 # Replaces Operator version
 # Set this when when there is a new patch release in the channel.
-REPLACES_VERSION ?= 0.2.6
+REPLACES_VERSION ?= $(shell cat REPLACES_VERSION)
```
Comment (yiraeChristineKim) [inline Makefile:9]: This too default

File: Makefile (hunk 1 above)
Comment (yiraeChristineKim) [reply Makefile:6]: ohohohoho nono it is good I just started review I couldn't see that

Comment (yiraeChristineKim) [reply Makefile:6]: can you write readme `docs/upgrading-gatekeeper.md`

File: Makefile [hunk 3]
```diff
@@ -95,12 +92,21 @@ all: build
...
+.PHONY: clean
+clean: IMG = quay.io/gatekeeper/gatekeeper-operator:v$(VERSION)
+clean: clean-manifests ## Clean up build artifacts.
 	rm $(LOCAL_BIN)/*
 
+.PHONY: clean-manifests
+clean-manifests: IMG = quay.io/gatekeeper/gatekeeper-operator:v$(VERSION)
+clean-manifests: kustomize unpatch-deployment bundle
+	# Reset all kustomization.yaml files
+	git restore **/kustomization.yaml
```
Comment (yiraeChristineKim) [inline Makefile:108]: lovely

File: config/olm-install/kustomization.yaml [hunk 4]
```diff
@@ -1,2 +1,27 @@
...
+      image: quay.io/gatekeeper/gatekeeper-operator-bundle-index:latest
+  target:
+    kind: CatalogSource
+    name: gatekeeper-operator
+- patch: |-
+    kind: Subscription
+    metadata:
+      name: gatekeeper-operator-sub
+    spec:
+      channel: "3.11"
```
Comment (yiraeChristineKim) [inline config/olm-install/kustomization.yaml:24]: Do we have `sed` for` channel: "3.11"`

**Comment by yiraeChristineKim** [comment]
Great! Thanks Dale!

File: Makefile (hunk 1 above)
Comment (dhaiducek) [reply Makefile:6]: There's a `release.md` doc that was also updated in this PR.

File: config/olm-install/kustomization.yaml (hunk 4 above)
Comment (dhaiducek) [reply config/olm-install/kustomization.yaml:24]: Yes--it's in the `deploy-using-olm` target and was updated to change this `kustomization.yaml` file in this PR.

**Review by yiraeChristineKim — DISMISSED** [review]

File: controllers/merge/merge.go [hunk 5]
```diff
@@ -62,6 +64,19 @@ func retainServiceFields(desiredObj, clusterObj *unstructured.Unstructured) erro
 	return nil
 }
 
+func retainServiceAccountFields(desiredObj, clusterObj *unstructured.Unstructured) error {
```
Comment (mprahl) [inline controllers/merge/merge.go:67]: Could you please add a test for this? A unit test should be fine.

File: config/olm-install/kustomization.yaml (hunk 4 above)
Comment (mprahl) [reply config/olm-install/kustomization.yaml:24]: Why not set it to `stable` by default?

File: scripts/kind-with-registry.sh [hunk 6]
```diff
@@ -4,20 +4,32 @@ set -o errexit
 # desired cluster name; default is "kind"
```
[the remark this answers is not in the document]
Comment (dhaiducek) [reply scripts/kind-with-registry.sh:4]: Good question--we should have something that cleans up both the cluster and the registry--I'll take a look.

File: config/olm-install/kustomization.yaml (hunk 4 above)
Comment (dhaiducek) [reply config/olm-install/kustomization.yaml:24]: Works for me! I was just taking the current value from the YAML.

**Comment by dhaiducek** [comment]
Updates:
- Added unit test for the ServiceAccount merging logic
- Added a `delete-test-cluster` target that deletes the Kind cluster and stops/removes the registry container/image
- Remove the unit tests as a prerequisite for `docker-build`
- Changed `docker-build` to use environment variables (minor thing, but it'll make it easier to revert if needed in the future)

**Review by mprahl — APPROVED** [review]

**Comment by yiraeChristineKim** [comment]
oh wait 5 commits?

**Comment by dhaiducek** [comment]
> oh wait 5 commits?

Yes--each is a separate change. And we probably only want to cherry-pick the bug fix, so separating the commits makes sense from that standpoint also.

**Pull request merged**